Reimplement OpenBabel::OBRandom as a wrapper of std::mt19937_64

e-kwsm opened the pull request:
Mersenne twister is one of high-quality PRNGs and available on all platforms since C++11.

e-kwsm committed:
feat: implement OpenBabel::OBRandomMT as a wrapper of std::mt19937_64

deprecates OpenBabel::OBRandom

ghutchis commented:
Looks like a good idea.

One thing that's been requested to add an OBOp plugin to set the RNG seed (e.g., #1934) - would you be willing to add that too.

e-kwsm committed:
refactor: switch to OBRandomMT if OB_USE_OBRANDOMMT is defined as truthy

the macro defaults to whether OpenBabel 4 is released or not

e-kwsm committed:
refactor: use std::unique_ptr for OpenBabel::OBConformerSearch::d

baoilleach commented:
I've said it before but we should be working to remove all uses of random numbers in the library.

ghutchis commented:
@baoilleach - I don't think that's possible. It's good to minimize them, allow a defined seed to be set, etc. but stochastic methods are critical to several tasks in the library.

ghutchis commented:
We've generally said that only major version changes (e.g., OB2 => OB3 or OB3 => OB4) would drop functions / methods.

The general concept is great, but can we either keep the existing code for backwards compatibility and *add* a new OBRandomMT class to use in conformer search, etc.? The other alternative would be to replace with the current code but wrap previous methods for backwards compatibility.

Otherwise we have to leave this for a while.

e-kwsm committed:
chore: remove unnecessary forward declaration

e-kwsm committed:
feat: add PRNG-related methods to OpenBabel::OBRandom

- UniformInt
- UniformReal
- Normal
- Bernoulli

e-kwsm committed:
refactor: use newly-added OpenBabel::OBRandom methods

e-kwsm committed:
refactor: improve method to generate 3-D random unit vector

this is enabled if OB_USE_IMPROVED_RANDOM_UNIT_VECTOR, which defaults to
whether OpenBabel 4 is released or not, is defined as truthy

e-kwsm committed:
chore: do not use pointer unnecessarily

e-kwsm committed:
docs: add explanations of OB_RANDOM_SEED environment variable

e-kwsm commented:
Done.

At the moment, i.e. `OB_VERSION < OB_VERSION_CHECK(4, 0, 0)`, the codes still use `OBRandom` and thus are backward compatible (rounding error may occur).
They will be automatically changed to utilize `OBRandomMT` when version 4 is released.
Explanations of `OB_RANDOM_SEED` environment variable in man docs are currently commented out.

alexander-telepov commented:
@e-kwsm thanks for such opportunity to fix seeds for reproducibility! However after build this pr I still can't repeat results both putting argument in obabel call from bash and after setting environment variable, for example for this CC1=CC(CNC2=CC=C(NC3CCC4=CC=C(C5=NNN=N5)C=C43)C=C2NC2=CC=C(F)N=C2F)=NC(C)=N1 mol. Can you please suggest some workaround?

e-kwsm commented:
@alexander-telepov You have to build Open Babel with `-DOB_USE_OBRANDOMMT` to enable the feature.

alexander-telepov commented:
@e-kwsm Thanks for such fast reply!

ghutchis commented:
@e-kwsm - thanks, I will get to this soon. I know you're updating all your branches - can you hold off a bit. Right now, the tests remain queued because GitHub limits the number of actions (e.g., a lot of branches are waking up right now).

e-kwsm commented:
> can you hold off a bit. Right now, the tests remain queued because GitHub limits the number of actions (e.g., a lot of branches are waking up right now).

Alright.

ghutchis committed:
Merge branch 'master' into mt19937_64

ghutchis commented:
While we're not going to remove the older OBRandom in 3.2, perhaps OBRandomMT should be the default compile option?

e-kwsm commented on src/config.h.cmake:
```diff
@@ -203,3 +203,8 @@
+
+// use OBRandomMT if true, otherwise deprecated OBRandom
+#ifndef OB_USE_OBRANDOMMT
+#define OB_USE_OBRANDOMMT (OB_VERSION >= OB_VERSION_CHECK(4, 0, 0))
```
> While we're not going to remove the older OBRandom in 3.2, perhaps OBRandomMT should be the default compile option?


```suggestion
#define OB_USE_OBRANDOMMT (OB_VERSION >= OB_VERSION_CHECK(3, 2, 0))
```

will enable the feature by default in 3.2.

ghutchis replied:
I think we should just switch the default in HEAD when this merges so there's some time to test, check fuzz, etc.

ghutchis replied:
I pushed a suggestion - adding a USE_RANDOMMT cmake flag that defaults to ON - but can be compiled off if people prefer the older RNG.

We won't toss the older OBRandom, but 3.2 will bump the ABI so it should work - and allow specifying seeds.

ghutchis replied:
Let me know if you're okay with this or would prefer a different approach.

e-kwsm replied:
LGTM :+1:

ghutchis committed:
Merge branch 'master' of github.com:openbabel/openbabel into mt19937_64

ghutchis committed:
Default to USE_RANDOMMT and add as a CMake flag to turn off

Signed-off-by: Geoff Hutchison <geoff.hutchison@gmail.com>

ghutchis commented:
@baoilleach - still some random numbers but now they can be seeded for reproducibility
